### ISSUE TITLE
Enhancement Ideas

### ISSUE THREAD

**Issue by chrishoage** [koekeishiya/kwm#42]
I would like to state again how blown away I am with this project. By far the best window manager I have used. 

The following are things that I have been looking for from window managers on OS X but have not been able to find in a single manager. 
They are merely suggestions and ideas, not necessarily things that should be be implemented immediately. Any examples are just to get the point across, not necessarily the configuration that should be used.

~~**Focus screen:** Unless I have missed something there is no way to cycle though focus onto another monitor. One must use the mouse to give focus to the other monitor. It would be convenient to be able to have kwmc to give another screen focus.~~ Implemented in:  https://github.com/koekeishiya/kwm/commit/df455f3505562e0ece42b5c782729e0df97012ac

```
kwmc bind ctrl+shift-1 screen -f 0
kwmc bind ctrl+shift-2 screen -f 1
kwmc bind ctrl+shift-p screen -f prev
kwmc bind ctrl+shift-n screen -f next
```

~~**Cycle focus options:** following the focusing screen, optionally being able to have the focus cycle to the other end would be nice. Maybe even further options to limit the focus cycle to one screen or all screens?~~ Implemented in: https://github.com/koekeishiya/kwm/commit/df455f3505562e0ece42b5c782729e0df97012ac

```
# Do not cycle focus (same as no config setting)
kwmc config cycle-focus disabled

# Cycle focus though all windows in screen
kwmc config cycle-focus screen

# Cycle focus though all windows in all screens
kwmc config cycle-focus all
```

~~**Mouse follows focus:** This would optionally move the mouse to the center of what ever window gains focused when using keyboard focus navigation.~~ Implemented in https://github.com/koekeishiya/kwm/commit/0ea0ae6294b6ca5cc1fa0e7209a8ca58a28a15f4

```
kwmc config focus mouse-follows enable
```

**Configure screen with layout:** this would allow for example one screen to be `bsp` and the other `monocle`. This would override what is set with `kwmc config space` for that screen. 

```
kwmc config screen 0 space bsp
kwmc config screen 1 space monocle
```

~~**Capture application to screen:** allow for applications/windows to be opened on a specific monitor if available.~~ Implemented in: https://github.com/koekeishiya/kwm/commit/9089ad64bc6ef707fc1e9062e9f71a82ce40569c 

```
kwmc config capture id application

kwmc config capture 1 iTunes
kwmc config capture 1 Slack

The new position always uses the normal logic, regardless of the
screen having focus or not. If the screen does not exist, capture
will be ignored.
```

This config would put new windows from the application in the first position of the specified screen if the screen did not have focus. If the screen did have focus the new window position is chosen by the normal logic. If the screen is not available it reverts back to normal (focus based) window creation.  

~~**Prefix key trigger:** There are a _ton_ of different key combinations on OSX, very few of them are configurable. Something I have been searching for is a WM on osx that allows for a tmux style prefix to capture commands. This is something that I have not really seen on OSX outside of shell applications. I have no knowledge of the difficulty of this, so for all I know this could be impossible on OSX and that is why it is so rare. But having the ability to prefix any commands might help with that.~~

Implemented in https://github.com/koekeishiya/kwm/commit/c27aee4034ec5345a207a013fe297c6bdf8395b0

Edit by Koe:
```
Bind specific prefixes is something I will not be doing due to needing a
more complex system than I want to deal with for Kwm.
A single global prefix however is something I would be interested in doing.
```
```
# Global Prefix
kwmc config prefix ctrl-k

# prefix (ctrl+k) then the l or h key with no mods 
kwmc bind -h window -f prev
kwmc bind -l window -f next


```

Again, I am not suggesting that all of these should be implemented right now, or even ever (they might not be good ideas for any number of reasons). It could be incredibly difficult to accomplish some (or all) of these and it might not be worth doing. Just curious of your thoughts on them and if you think any of them are a good idea.

Thanks!

**Comment by koekeishiya** [comment]
Most of the ideas seem like they would be very nice to have, and some of them seems quite tricky to implement, however I'll keep it in mind for future changes that will be made.

Regarding the changes that affect the use of multiple monitors, a lot of the current system in Kwm will probably be reworked. I've just not had any monitors to do testing with during christmas/new years.

Personally I do not find mouse-follows-focus useful, and it could potentially bring in some hard-to-fix bugs caused by delay (what if you change focus 3times in a row; instantly), so I'm curious to the use case of this?

**Comment by flou** [comment]
Prefix key trigger would indeed be awesome!
It may not be the exact same thing but Sublime Text does use a prefix key (for instance: `cmd+k, cmd+u` for converting to upper case), but it is limited to the application.

**Comment by chrishoage** [comment]
> Personally I do not find mouse-follows-focus useful, and it could potentially bring in some hard-to-fix bugs caused by delay (what if you change focus 3times in a row; instantly), so I'm curious to the use case of this?

Pure laziness!  90% of the time I want my mouse over the window that is focused. It was (is?) a feature of Amethyst that I kinda sorta miss. It's on my wish list, but by no means important. I wouldn't be sad if it was never implemented.  

> Regarding Capture Application to screen, I really do like this idea. At the current time, I'm not sure if it's doable tho. However, would it capture ALL windows from that application, or just the next one to be opened?

I imagine it would be any window that a specified application generates for predicability and consistency. In my head I'm imagining a "catch all" config option. For example i always have my laptop display (when I'm connected to my main monitor) be in `monocle` mode with Slack and iTunes both on that display. Having this behavior happen via config would be super slick. 

 > Regarding the changes that affect the use of multiple monitors, a lot of the current system in Kwm will probably be reworked. I've just not had any monitors to do testing with during christmas/new years.

Do you not own a 2nd monitor or are you just in a location where you do not have access to a 2nd monitor? If it's the former consider starting a https://www.patreon.com/ because I know I would throw some money your way for the work you've put in (especially if you plan to keep working and improving it) 

 You've gained quite a few github stars in a fairly short amount of time. I bet with a little publicity again on reddit and hacker news you could hit any patreon goal you might want.


Some more ideas I had last night: 

**kwmrc config reload:** right now the only way for kwm to read new config settings in `.kwmrc` is to `killall kwm` and have the service restart. This is less than ideal because it loses the state of my tree and forces me to relay out the windows. 

```
kwmc config reload
```

**Save bsp tree:** This is a "pie in the sky" idea, and probably a _significant_ amount of work. But i'd really love a way to store a tree and apply it at will. This has a ton of questions: How would you serialize the tree? Would it be configurable in the kwmrc or a binary blob? 

My development window environment has a pretty specific layout (Two browser windows, a sublime text instance with a large chunk of the display, and small iTerm windows below it). Having a way to get back to that state via a kwmc command would be amazing.

**Comment by koekeishiya** [comment]
Basic mouse-follows-focus should be quite easy to implement and I can definitely look into doing this for version 1.0.4. I assume focus-follows-mouse would be disabled if mouse-follows-focus is enabled, and so you choose 1 out of those 2 modes.

Capture application to screen would be implemented (if possible) using a 'catch all' approach at first at least, and maybe it can be extended later on.

I do have quite a few monitors, just not at the location I'm currently at, so that's not a problem.
The reload config option is long overdue and will definitely happen sometime soon.
It is probably the most annoying thing I deal with during development at the moment.

I have already started thinking about being able to save bsp-sessions, however I ran into the problem where a running window can only be identified by a unique window number, and this window number is different every time. Because of this, I cannot properly restore a session that was saved, because I have no idea what the window number would be the next time these applications start.

I thought about matching using the application name and window title, however there may arise conflicts (e.g iTerm2 windows often have the same title if they are both a bash prompt).

**Comment by chrishoage** [comment]
> I assume focus-follows-mouse would be disabled if mouse-follows-focus is enabled, and so you choose 1 out of those 2 modes.

Actually I was imagining both would exist in conjunction? If it's a technical limitation for having both working I'd pick `focus-follows-mouse` any day of the week because it's way more useful than `mouse-follows-focus`. As I said i'm not hurting for `mouse-follows-focus`, it can be placed on the back burner (or never be implemented).

**Comment by koekeishiya** [comment]
Well they could work simultaneously, however every time a window would gain focus due to focus-follows-mouse,  the mouse-follows-focus would also kick in and move your mouse to the center of that window.

**Comment by chrishoage** [comment]
>  the mouse-follows-focus would also kick in and move your mouse to the center of that window.

The way I was envisioning it working is `mouse-follows-focus` would _only_ be triggered when using `kwmc window -f prev | next`. So i'd be apart of the command: First, move focus to the requested window, second move mouse to the center of that window if the config is set. 

It wouldn't happen in response to an `onFocus` event of a window. 

Again no idea how feasible any of it is, just ideas.  

> I cannot properly restore a session that was saved, as I have no idea what the window number would be the next time these applications start.

What about a middle ground: Save the layout, but ignore the windows. If a layout was restored only the tile position and sizes would be used, the apps/windows applied to those would be filled in after the fact and there is no guarantee that a specific window goes to a specific spot. 

This is not "perfect" but would still save a bit of time having to change all the sizes and positions of the tiles. It's pretty trivial to move the window to the correct spot as long as the layout is setup already.

**Comment by koekeishiya** [comment]
Alright so having mouse-follows-focus only trigger on `kwmc window -f prev | next` does make sense and is very easily implemented. This would allow for having both focus-follows-mouse and mouse-follows-focus to co-exist.

Being able to save the bsp-layout should be possible to do. I imagine that it would then fill in the containers (starting from left -> right), for each new window that is detected / opened after the restore command has been issued, until all containers are taken. Then the 'normal' bsp-tiling would take place for new windows.

**Comment by chrishoage** [comment]
>  I imagine that it would then fill in the containers (starting from left -> right), for each new window that is detected / opened after the restore command has been issued, until all containers are taken. Then the 'normal' bsp-tiling would take place for new windows.

Yeah this is pretty much how I was imagining it would work. Still open questions around how to manage them. What would the commands look like? Where would the configuration files be stored? Would they be serializable (and human editable) or binary? 

By the way if you need/want any help maintaining the docs please let me know anything you'd like done. I don't trust myself to write C/C++ or else I'd offer to help out that way. 

I have started crossing things off this list and referencing the commit to keep track. Feel free to cross things off that you don't wish to implement.

**Comment by koekeishiya** [comment]
For commands I would like to have something like `kwmc tree save name` and `kwmc tree restore name`. This would save/restore the layout for the active space of the focused monitor.

I assume that if this happens, we would move configs to ~/.config/kwm/.kwmrc and ~/.config/kwm/layouts/name.ext, or even just ~/.kwm/stuff. I'm not sure yet what the best serialization method would be, but preferably human readable and editable.

Regarding documentation, if you find something unclear or think that it could be explained better feel free to make such a change and submit a pull request or something.

I'm really bad at explaining things. In addition to this I've tried to keep the documentation short and concise as I get the impression that people are not gonna bother reading 12pages of information.

**Comment by koekeishiya** [comment]
I've added support for a global prefix trigger, however currently it will only respond to one command before the prefix-key has to be hit once more.

I do plan on implementing a timing system to allow for multiple keypresses. Basically how I imagine this to work is that a timer is started when the prefix key has been hit (lets use 5seconds as an example).
The user now has a 5second window to hit a Kwm hotkey, and for each hotkey he presses, the timer will be extended by an additional 2seconds. When the timer reaches 0, the status of the prefix key resets.
Thoughts?

**Comment by chrishoage** [comment]
The timing system you describe sounds pretty good. 

I was curious your thoughts on bind specific prefixes? 

My use case is this: Commands like moving focus or moving a window I would like to keep quick key bindings. Bindings like changing a spaces setting from `bsp` to `monocle` or a toggling a window floating I don't use particularly often. Burying those in a prefix will allow for fewer conflicts and not have much effect on workflow.

However the timing system you describe may alleviate that issue. 

The documentation is actually really good, I do have some thoughts around making it easier to install/setup. I will open separate ticket for ideas there.

**Comment by koekeishiya** [comment]
I did an edit of the main post, I probably should've added in my own as well.

```
Bind specific prefixes is something I will not be doing due to needing a
more complex system than I want to deal with for Kwm.
A single global prefix however is something I would be interested in doing.
```

**Comment by koekeishiya** [comment]
Right so I think I'm gonna have this be the final hotkeys system for now. Bind-specific prefixes would be an extension of this system (not very likely to happen for a long time, if ever, don't get your hopes up).

The next feature that I personally want to look into next, is being able to save and restore bsp-sessions. Whatever possible to get out of it anyway. Pretty certain that the commands will be 
`kwmc tree save name` and `kwmc tree restore name` as stated above.

I will post more information about my thought process on this as I work through pros and cons to different ways of implementing this.

**Comment by chrishoage** [comment]
>  Bind-specific prefixes would be an extension of this system (not very likely to happen for a long time, if ever, don't get your hopes up).

That is fine, I think now that prefix-key works with a timeout it would help with the usability. Allowing no modifiers when prefix is set is a welcome addition.  

Looking forward to saving the tree!

**Comment by koekeishiya** [comment]
Regarding bind-specific-prefix, I have worked out a way to implement it, so that is not the problem. I just feel like it's no point making it more complex than necessary.

Hopefully the prefix-key works quite well with the current timeout. It works like the following:
After the prefix-key has been hit, the user has 2seconds to hit a Kwm hotkey, before the prefix-state resets. If a valid hotkey is pressed during this time, it will refresh back up to 2seconds.

During my testing, even the 2second window felt long when I knew I was finished doing whatever actions I wanted, and it would interfere with my regular typing, but I guess this heavily depends on the user and how accustomed they are to their hotkeys. By hitting the prefix-key before the 2second window is up, the prefix-state will be turned off.

**Comment by chrishoage** [comment]
> During my testing, even the 2second window felt long when I knew I was finished doing whatever actions I wanted, and it would interfere with my regular typing, but I guess this heavily depends on the user and how accustomed they are to their hotkeys.

This is why I suggested both bind specific prefix as well as global prefix. Kmwc commands like changing padding, gaps, toggling a window float, changing a space layout style (bsp -> monocle), filling a window to its parent all (for me at least) could be buried behind a prefix because they are actions that do not happen often, and when they do typically take time. While I agree bind specific prefixes would make the bind system more complex, it would only make it complex for people who wanted more power. 

Right now the kwm defaults conflict with some Sublime Text and iTerm key commands. To me it would be easier to remember `ctrl+k f` to toggle something to float than to `ctrl+alt+shift+f` or something, and then worrying about conflicts with any other application.  

Moving focus or windows is something that happens so often I would not want these commands buried under a prefix but others I can see real value in it. 

Out of curiosity what do you think about a configuration option for the prefix timeout? something like 

```
# Configure how long a prefix should remain active for in seconds.
# 0 disables the prefix after the first bound command
kwmc config prefix-timeout 0.5
```

I have not had the chance to play with the prefix since the timeout has been added. Maybe my thoughts would change after I get the chance later today.

**Comment by koekeishiya** [comment]
That is a lot more keybinds overlapping than i originally imagined. I personally only really have vim, tmux and some other keybinds that I've set up myself, so its not that bad from my perspective.

I have indeed added an option to specify the prefix-timeout now. After using this some more, I feel like having a 0.75 second window was the most effective for me. I can get my things done, and it also allows me to naturally transition out. Again this is something that I think will differ greatly between different users.

I am very curious about how this works for other people, and if it turns out to be "less valuable" than expected, I could give a shot at supporting bind-specific-prefixes.

**Comment by chrishoage** [comment]
After using only gobal prefixes for a while I have to say the navigation commands I wish were not prefixed. 

```
kwmc config prefix ctrl-k
kwmc config prefix-timeout 0.75
kwmc bind -f window -t float
kwmc bind ctrl-f space -t float
```

These types of commands are great to have behind a prefix because it's easy to associate the modifiers with the command. I can reuse `f` with out worrying about conflicts with other apps. 

Navigation commands feel better with out a prefix because they happen so much:

```
# Move Focus
kwmc bind cmd+alt-h window -f prev
kwmc bind cmd+alt-l window -f next

# Move Window
kwmc bind cmd+alt-p window -s prev
kwmc bind cmd+alt-n window -s next
```

I can imagine how having to handle a bunch of different prefixes would make things quite complex. How about an alternative syntax to limit the prefixes that are kept track of to 1.

```
kwmc config prefix ctrl-k
kwmc config prefix-global disable #disable is default. Just to show the prefix can be applied globally
kwmc config prefix-timeout 0.75

kwmc bind prefix-f window -t float
kwmc bind prefix-ctrl-f space -t float

 # Move Focus
kwmc bind cmd+alt-h window -f prev
kwmc bind cmd+alt-l window -f next

# Move Window
kwmc bind cmd+alt-p window -s prev
kwmc bind cmd+alt-n window -s next
```

I think I have found a decently happy equilibrium with out prefixes but I would love some only commands prefixed to make it easier to remember.

**Comment by koekeishiya** [comment]
I edited your main post to reflect the new changes.
I've only tested them briefly but it appears to work properly.

**Comment by chrishoage** [comment]
```
kwmc screen -c 1 iTunes
kwmc screen -c 1 Slack
```

Does not appear to work for me

I also tried 

```
kwmc config screen -c 1 iTunes
kwmc config screen -c 1 Slack
```

Incase it had was typo. 

```
kwmc config cycle-focus screen | all
```

Does not appear to work (at least when there are two monitors connected. Have not tried with just one yet.

**Comment by chrishoage** [comment]
Gah, for some reason 
`make install` then `killall kwm` did not actually update the binary. I had to `rm -rf ./bin` then `make install` before the changes would pick up. 

`cycle-focus` and `screen -f` both now function, but I still can't get screen capture to work

**Comment by koekeishiya** [comment]
´kwmc screen -c 1 iTunes` works fine for me. My macbook display has index 0 and the external monitor is 1, also just tested with Finder and Mail as well. Both when reading from config or applying it through a terminal while Kwm is running

**Comment by chrishoage** [comment]
Odd, ok now it is working. However sometimes when `screen -m 1` it does not update the display that the window moved to correctly. It requires me to force what ever tiling mode that was there again. This does not happen consistently though. 

Also, rarely _all_ window positions get screwed up, is there a way to run `window -c refresh` for each and every window vs having to focus each one and run that command? 

Another issue is when using `screen -f prev | next` _sometimes_ the display that gains focus does not give any window focus. It appears that the new display gets focus, a window gets focus for a second, then no window has focus at all. It is easily fixed by just running the screen focus command again. 

I have run into more keybinding conflicts with applications. Are command specific prefixes something I could easily add myself in `hotkeys.cpp`? If so could you provide an example for how I might do that? I would greatly appreciate _some_ way to achieve command specific prefixes even if it can't be in the `kwmrc`. Thanks!

**Comment by koekeishiya** [comment]
Uh, the reasoning for using `screen -c 1` is basically c == capture, although I could probably change it up.
I'll look closer into the focus issues, been using my android tablet as a second monitor when working on that code so it was amazingly annoying to perform testing.

As for your question, you can indeed implement that logic in hotkeys.cpp.
What you would need to do is more or less the following:

make a struct or just bools for the state of the "hotkey" that will be your pefix
declare global bool PrefixActivated;

In the KWMHotkeyCommands function, the very first thing that happens, i that you have to match the given mods and keycode (Mod.CmdKey etc, Keycode) etc for your prefix, if it matches the prefix, set PrefixActivated to true, and also return true from the hotkey function at this point.

Now the next time this function is run (aka next keypress you do), the PrefixActivated will be true, and you can now perform normal key-check where you set KwmcCommandToExecute to whatever command you want to run (do not call return implicitly here). This should be all you need for a single prefix for some binds.

Multiple prefixes will require some more thinking than the above.

**Comment by chrishoage** [comment]
> Uh, the reasoning for using screen -c 1 is basically c == capture, although I could probably change it.

Ah, what I was saying is it seems odd it's not `kwmc config screen -c 1 iTunes` as it seems like it is a config option rather than a command?

I don't need multiple prefixes, I will give your instructions a shot thank you. I do hope you reconsider bind specific prefixes (I don't think you should worry about multiple prefixes ever), there are just _so_ many key binds in other applications it's pretty difficult to find easy to remember key combos that don't conflict. 

Basically want I want/need is Window/Screen focus/move commands un-prefixed and everything else prefixed. It would be ideal for easy of memorization and lack of conflicts. 

Thanks again!

**Comment by koekeishiya** [comment]
I can potentially allow a single prefix that is bind specific (togge global) using the `kwmc bind` syntax yeah. If you need more help implementing it yourself in hotkeys.cpp for the time being, let me know.

Does `kwmc config capture screenid application` seem better than the current one?

**Comment by chrishoage** [comment]
Yeah your example screen capture command makes it seem more like a config
option than a bind command.

A single bind specific prefix would (in my opinion) strike the perfect
balance of simplicity and power with a toggle for global.

I will try not to bother you about my hotkeys.cpp though, especially if
you're considering the above. I will keep your offer in mind if I just
can't figure it out from the existing code.

Thank you very much for everything!

**Comment by koekeishiya** [comment]
"Also, rarely all window positions get screwed up, is there a way to run window -c refresh for each and every window vs having to focus each one and run that command? "

I have narrowed this down to happening once every time an external monitor is connected and the mouse is used to focus a different monitor. If the mouse-movement does not land the cursor on a window on the new monitor, this bug will happen. Again, it is only once every time a new monitor is connected. I have tried to debug this, and even tho I know exactly how it happens, I have yet to discover why.

Regarding `screen -m stuff`, what happens ( is supposed to happen ) is that the window will un-tile from its position and sent to a new monitor. It will tile using the "startup logic" that Kwm uses for bsp, however the other windows on that screen will not resize until you focus the monitor (The window-tree has been properly modified, the changes will just not be 'applied' until focus is given). This is due to a technical limitation with how OSX report windows and how Kwm fundamentally handles this in response.

**Comment by chrishoage** [comment]
> Moving a window from a monocle-screen to a bsp-screen and vice versa is probably not working at the moment and could potentially break things.

This is exactly the issue. 

>  Rather than going for the manual fix you requested, I want to try and make this not happen because it's a pretty stupid bug to have around.

Sounds good to me. 

Since nearly all of the requests in this ticket have been addressed do you mind if I close this ticket and open new more focused tickets to track the renaming items and bugs listed?

**Comment by koekeishiya** [comment]
Sure, that sounds like a good idea.
